data_structures/valid_parens.py

```diff
@@ -0,0 +1,25 @@
+class Solution:
+    def __init__(self):
+        self.stack = []
+
+    def isValid(self, s: str):
+        if not len(s):
+            return True
+        if len(s) % 2:
+            return False
+        for x in s:
+            if not len(self.stack):
+                self.stack.append(x)
+            else:
+                prev_idx = len(self.stack) - 1
```

Comment:
correct, but can be done easier - `self.stack[-1]`

Reply:
it throws an exception..

Reply:
sorry, I've didn't explain it properly. `stack[-1]` already takes an element, so it would fail, if stack is empty. However, if you encounter closing parentheses and stack is empty - you don't need to remove anything, you need just to stop and return `False`